lib/pbench/server/api/resources/query_apis/search_api.py

```diff
@@ -0,0 +1,224 @@
+                    if type(element) is dict:
+                        recurse(element, prefix + ".", out)
+                    else:
+                        recurse(element, prefix + str(i) + ".", out)
+            else:
+                key = prefix[:-1]
+                if key not in out:
+                    out[key] = es_json
+                elif type(out[key]) == list:
+                    out[key].append(es_json)
+                else:
+                    out[key] = [out[key], es_json]
+            return out
+
+        return recurse(es_json, prefix="", out={})
+
+    def assemble(
+        self, json_data: Dict[AnyStr, Any], context: CONTEXT
+    ) -> Dict[AnyStr, Any]:
+        """
+        Construct a pbench search query based on a pattern matching given "search_term" parameter
+        within a specified date range and which are either owned by a specified username,
+        or have been made publicly accessible.
+
+        {
+            "user": "username",
+            "start": "start-time",
+            "end": "end-time"
+            "search_term": "string_pattern_to_search"
+            "fields": list(interested run document fields returned from IndexMapping API)
+        }
+
+        json_data: JSON dictionary of type-normalized parameters
+            user: specifies the owner of the data to be searched; it need not
+                necessarily be the user represented by the session token
+                header, assuming the session user is authorized to view "user"s
+                data. If "user": None is specified, then only public datasets
+                will be returned.
+
+            "start" and "end" are datetime objects representing a set of Elasticsearch
+                run document indices in which to search.
+
+            "search_term": string pattern to match against all the datasets owned by a specified
+                    username within a given time range
+
+            "fields": List of run document fields end user is interested in. E.g.
+                    ["@timestamp", "run.controller", "run.name", ...]
+                    The fields controls what fields are returned by the query, and it has no effect
+                    on which fields are searched.
+        """
+        user = json_data.get("user")
+        start = json_data.get("start")
+        end = json_data.get("end")
+
+        # If the search_term parameter is not specified we will return all the Elasticsearch hits
+        search_term = json_data.get("search_term", "")
+        # If no fields are specified, the query will return all the fields from Elasticsearch hits
+        selected_fields = json_data.get("fields", [])
+
+        # We need to pass string dates as part of the Elasticsearch query; we
+        # use the unconverted strings passed by the caller rather than the
+        # adjusted and normalized datetime objects for this.
+        start_arg = f"{start:%Y-%m}"
+        end_arg = f"{end:%Y-%m}"
```

Comment:
Why only `YYYY-MM` here? I'm not sure how Elasticsearch will default the day on the range search, but could this include days at the beginning we don't want and exclude days at the end we might want?